framework/helpers/docker.py

```diff
@@ -0,0 +1,308 @@
+        )
+        destination = self.mount_dir() / "bootstrap.json"
+        with open(destination, "w") as f:
+            f.write(file)
+            logging.debug(f"Generated bootstrap file at %s", destination)
+
+    def _MakeWorkingDir(self, base: str):
+        for i in range(100):
+            # Date time to string
+            run_id = datetime.now().strftime("%Y%m%d_%H%M%S")
+            self.__mount_dir = base / f"testrun_{run_id}"
+            if not self.__mount_dir.exists():
+                logging.debug(f"Creating %s", self.__mount_dir)
+                self.mount_dir().mkdir(parents=True)
+                return
+        raise Exception("Couldn't find a free working directory")
+
+    def mount_dir(self) -> Path:
+        if self.__mount_dir is None:
+            raise RuntimeError("Working dir was not created yet")
+        return self.__mount_dir.absolute()
+
+    def xds_config_server_port(self, n: int):
+        return self.__ports[n]
+
+
+class ChildProcessEvent:
+    def __init__(self, source: str, data: str):
+        self.source = source
+        self.data = data
+
+    def __str__(self) -> str:
+        return f"{self.data}, {self.source}"
+
+    def __repr__(self) -> str:
+        return f"data={self.data}, source={self.source}"
+
+
+class ProcessManager:
+    def __init__(
+        self,
+        bootstrap: Bootstrap,
+        nodeId: str,
+        verbosity="info",
+    ):
+        self.dockerClient = DockerClient.from_env()
+        self.nodeId = nodeId
+        self.__outputs = {}
+        self.__queue = Queue()
+        self.__bootstrap = bootstrap
+        self.verbosity = verbosity
+
+    def NextEvent(self, timeout: int) -> ChildProcessEvent:
+        event: ChildProcessEvent = self.__queue.get(timeout=timeout)
+        source = event.source
+        message = event.data
+        logging.info(f"[%s] %s", source, message)
+        if not source in self.__outputs:
+            self.__outputs[source] = []
+        self.__outputs[source].append(message)
+        return event
+
+    def ExpectOutput(self, source: str, message: str, timeout_s=5) -> bool:
+        logging.debug(f'Waiting for message "%s" on %s', message, source)
+        if source in self.__outputs:
+            for m in self.__outputs[source]:
+                if m.find(message) >= 0:
+                    return True
+        deadline = datetime.now() + timedelta(seconds=timeout_s)
+        while datetime.now() <= deadline:
+            event = self.NextEvent(timeout_s)
+            if event.source == source and event.data.find(message) >= 0:
+                return True
+        return False
+
+    def OnMessage(self, source: str, message: str):
+        self.__queue.put(ChildProcessEvent(source, message))
+
+    def mount_dir(self):
+        return self.__bootstrap.mount_dir()
+
+
+def _Sanitize(l: str) -> str:
+    if l.find("\0") < 0:
+        return l
+    return l.replace("\0", "�")
```

Comment:
What's `�`?

Reply:
Just a placeholder. Need to find why our code logs \0. I opened a Core bug and assigned to myself.

Reply:
Can you link the bug here please? Was it resolved?